### PR TITLE
[SYCLomatic] Fix migration bug of `__attribute__(__noinline__)`.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -279,6 +289,12 @@ void IncludesCallbacks::MacroDefined(const Token &MacroNameTok,
     if (!II)
       continue;
 
+    // The "__noinline__" macro is re-defined by CUDA compiler.
+    // When it is used in "__attribute__()", we cannot replace it.
+    if (II->hasMacroDefinition() && (II->getName().str() == "__noinline__")) {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
    if (II->hasMacroDefinition() && (II->getName() == "__noinline__")) {
```

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 1 above)
Comment (tangjj11) [reply clang/lib/DPCT/ASTTraversal.cpp]: fixed

**Review by yihwang-nv — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]

**Review by ziranzha — APPROVED** [review]

File: clang/lib/Parse/ParseDecl.cpp [hunk 2]
```diff
@@ -194,6 +194,8 @@ void Parser::ParseGNUAttributes(ParsedAttributes &Attrs,
   SourceLocation StartLoc = Tok.getLocation();
   SourceLocation EndLoc = StartLoc;
 
+  llvm::SaveAndRestore<bool> SART(PP.IsInAttr, true);
```
Comment (zhimingwang36) [inline clang/lib/Parse/ParseDecl.cpp:199]: add #SYCLomatic_CUSTOMIZATION to protect it?

File: clang/include/clang/Lex/Preprocessor.h [hunk 3]
```diff
@@ -3020,6 +3020,10 @@ class Preprocessor {
   bool setDeserializedSafeBufferOptOutMap(
       const SmallVectorImpl<SourceLocation> &SrcLocSeqs);
 
+#ifdef SYCLomatic_CUSTOMIZATION
+  bool IsInAttr = false;
```
Comment (zhimingwang36) [inline clang/include/clang/Lex/Preprocessor.h:3024]: have not found this value set to other value?

File: clang/lib/Parse/ParseDecl.cpp (hunk 2 above)
Comment (zhimingwang36) [reply clang/lib/Parse/ParseDecl.cpp:199]: why here set the value to true,  could you add comment?

Comment (tangjj11) [reply clang/lib/Parse/ParseDecl.cpp:199]: added.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**